### PR TITLE
chore(web): Refactor auth logic

### PR DESCRIPTION
**Pull request by SokratisVidros**
### What changed? Why was the change needed?

1. Introduce useAuth() to simplify auth state management for the web-app. This work is necessary to support multiple auth providers for SSO features.
2. Apply a minor refactoring to LaunchDarkly handling, aiming for a simple implementation in the next commit.

### CODE REVIEW

File: apps/web/src/hooks/useBlueprint.ts [hunk 1]
```diff
@@ -12,10 +12,10 @@ export const useBlueprint = () => {
   const { pathname } = useLocation();
   const segment = useSegment();
   const id = localStorage.getItem('blueprintId');
-  const { token } = useAuthController();
+  const { token } = useAuth();
 
   useEffect(() => {
-    if (id && !!token) {
+    if (id) {
```
Comment (antonjoel82) [inline apps/web/src/hooks/useBlueprint.ts:18]: ❓ **question:** Should we remove `token` from this altogether?‏

File: apps/web/src/pages/auth/InvitationPage.tsx [hunk 2]
```diff
@@ -8,15 +8,16 @@ import { getInviteTokenData } from '../../api/invitation';
...
 import { useAcceptInvite } from './components/useAcceptInvite';
 import { LoginForm } from './components/LoginForm';
 
 export default function InvitationPage() {
   const queryClient = useQueryClient();
   const navigate = useNavigate();
-  const { token, logout, currentUser } = useAuthContext();
+  const { token, logout, currentUser } = useAuth();
   const location = useLocation();
+  // TODO: Replace token check with currentUser check
```
Comment (antonjoel82) [inline apps/web/src/pages/auth/InvitationPage.tsx:20]: ❓ **question:** What does this mean? Could we do this now?‏

File: apps/web/src/pages/auth/components/SignUpForm.tsx [hunk 3]
```diff
@@ -62,15 +62,15 @@ export function SignUpForm({ invitationToken, email }: SignUpFormProps) {
      * To get the correct token when sending the request
      */
     const token = (response as any).token;
-    applyToken(token);
+    // login(token);
```
Comment (antonjoel82) [inline apps/web/src/pages/auth/components/SignUpForm.tsx:65]: ❓ **question:** Why is this commented? Do we need this?
‏

File: libs/shared/src/types/auth/auth.types.ts [hunk 4]
```diff
@@ -10,6 +10,9 @@ export interface IJwtPayload {
   exp: number;
 }
 
+// @deprecated Use IJwtClaims instead
```
Comment (antonjoel82) [inline libs/shared/src/types/auth/auth.types.ts:13]:
```suggestion
/**
* @deprecated Use IJwtClaims instead
*/
```

Comment (antonjoel82) [reply libs/shared/src/types/auth/auth.types.ts:13]: ❓ **question:** Why this naming change?‏

File: libs/shared-web/src/hooks/useFeatureFlags.ts
Comment (antonjoel82) [inline libs/shared-web/src/hooks/useFeatureFlags.ts:1]: 🤓 **nitpick (non-blocking):** Should rename this file to `useFeatureFlag.ts` then‏

File: libs/shared-web/src/hooks/useEnvController.ts [hunk 5]
```diff
@@ -64,19 +64,20 @@ export const useEnvController = (
       }
 
       setIsLoading(true);
+      // TODO: Do we need to get the token from this endpoint?
```
Comment (antonjoel82) [inline libs/shared-web/src/hooks/useEnvController.ts:67]: 🗒 **note (non-blocking):** This seems necessary based on Richard mentioning that we extract `environment` and `orgId` from the token‏

File: apps/web/src/Providers.tsx [hunk 6]
```diff
@@ -26,7 +24,7 @@ const queryClient = new QueryClient({
 });
 
 /** Full-page loader that uses color-preferences for background */
-const fallbackDisplay = (
+const Fallback = (
```
Comment (antonjoel82) [inline apps/web/src/Providers.tsx:27]: ❓ **question:** Can we delete this then?‏

File: libs/shared-web/src/constants/routes.enum.ts [hunk 7]
```diff
@@ -47,3 +47,10 @@ export enum ROUTES {
   SECURITY = '/settings/security',
   BILLING = '/settings/billing',
 }
+
+export const PUBLIC_ROUTES = [
```
Comment (antonjoel82) [inline libs/shared-web/src/constants/routes.enum.ts]:
```suggestion
export const PUBLIC_ROUTES = new Set<ROUTES>([
```

File: apps/web/src/Gate.tsx [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import { type ReactNode, useEffect, useLayoutEffect } from 'react';
+import { useLDClient } from 'launchdarkly-react-client-sdk';
+import { useAuth } from '@novu/shared-web';
+
+export function Gate({ children }: { children: ReactNode }) {
```
Comment (antonjoel82) [inline apps/web/src/Gate.tsx]: ✏ **suggestion:** Rename more semantically and add comment about the purpose fo the function‏

Comment (antonjoel82) [inline apps/web/src/Gate.tsx]:
```suggestion
export function Gate({ children }: PropsWithChildren<{}>) {
```

File: apps/web/src/Gate.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+        anonymous: true,
+      });
+    }
+  }, [ldClient, currentOrganization]);
+
+  useLayoutEffect(() => {
+    if (inPublicRoute || !isLoading) {
+      document.getElementById('loader')?.remove();
+    }
+  }, [inPublicRoute, isLoading]);
```
Comment (antonjoel82) [inline apps/web/src/Gate.tsx]: ✏ **suggestion:** IMO should be a separate hook‏

File: apps/web/public/index.html [hunk 10]
```diff
@@ -42,6 +42,27 @@
...
+          background-color: #EDF0F2;
+        }
+      }
+      #loader {
+          display: grid;
+          width: 100vw;
+          height: 100vh;
+          place-items: center;
+        }
+    </style>
```
Comment (antonjoel82) [inline apps/web/public/index.html]: ✏ **suggestion:** Prefer in different file for keeping this top-level file cleaner‏

File: apps/web/public/index.html [hunk 11]
```diff
@@ -42,6 +42,27 @@
...
+        }
+
+        body {
+          background-color: #EDF0F2;
+        }
+      }
+      #loader {
+          display: grid;
+          width: 100vw;
+          height: 100vh;
```
Comment (antonjoel82) [inline apps/web/public/index.html]:
```suggestion
          width: 100dvw;
          height: 100dvh;
```

File: apps/web/public/index.html [hunk 12]
```diff
@@ -42,6 +42,27 @@
...
+      @layer critical {
+        /* Override Mantine injected CSS layers to prevent minor jumps */
+        *,::before,::after,::backdrop,::file-selector-button {
+          margin: 0px;
+          padding: 0px;
+          box-sizing: border-box;
+        }
+
+        body {
+          background-color: #EDF0F2;
```
Comment (antonjoel82) [inline apps/web/public/index.html]:
```suggestion
          /* surface.page (light mode) */
          background-color: #EDF0F2;
```

File: apps/web/src/index.tsx [hunk 13]
```diff
@@ -1,19 +1,45 @@
...
+        kind: 'user',
+        anonymous: true,
+      },
+      options: {
+        // eslint-disable-next-line max-len
+        // https://docs.launchdarkly.com/sdk/features/bootstrapping#:~:text=Alternatively%2C%20you%20can%20bootstrap%20feature%20flags%20from%20local%20storage%3A
+        bootstrap: 'localStorage',
+      },
+    });
+  }
```
Comment (antonjoel82) [inline apps/web/src/index.tsx:32]: ✏ **suggestion:** Prefer this extracted to an aptly named file‏

File: libs/shared/src/types/auth/auth.types.ts (hunk 4 above)
Comment (SokratisVidros) [reply libs/shared/src/types/auth/auth.types.ts:13]: It's a standard naming convention when it comes to JWTs https://www.iana.org/assignments/jwt/jwt.xhtml#claims

File: libs/shared-web/src/constants/routes.enum.ts (hunk 7 above)
Comment (SokratisVidros) [reply libs/shared-web/src/constants/routes.enum.ts]: Done.

File: apps/web/src/Providers.tsx (hunk 6 above)
Comment (SokratisVidros) [reply apps/web/src/Providers.tsx:27]: Done

**Pull request merged**